src/librustc_mir/transform/simplify_cfg.rs

```diff
@@ -0,0 +1,121 @@
+
+        let mut worklist = vec![START_BLOCK];
+        while let Some(bb) = worklist.pop() {
+            for succ in mir.basic_block_data(bb).terminator.successors() {
+                if !seen[succ.index()] {
+                    seen[succ.index()] = true;
+                    worklist.push(*succ);
+                }
+            }
+        }
+
+        util::remove_basic_blocks(mir, &seen);
+    }
+
+    fn collapse_jump_chains(&self, mir: &mut Mir) {
+
+        // Find the target at the end of the jump chain, return None if there is a loop
+        fn final_target(mir: &Mir, mut target: BasicBlock) -> Option<BasicBlock> {
+            // Keep track of already seen blocks to detect loops
+            let mut seen: Vec<BasicBlock> = Vec::with_capacity(8);
+
+            while mir.basic_block_data(target).statements.is_empty() {
+                match mir.basic_block_data(target).terminator {
+                    Terminator::Goto { target: next } |
+                    Terminator::Panic { target: next } => {
+                        if seen.contains(&next) {
+                            return None;
+                        }
+                        seen.push(next);
+                        target = next;
+                    }
+                    _ => break
+                }
+            }
+
+            Some(target)
+        }
+
+        for bb in mir.all_basic_blocks() {
+            // Temporarily swap out the terminator we're modifying to keep borrowck happy
+            let mut terminator = Terminator::Diverge;
+            mem::swap(&mut terminator, &mut mir.basic_block_data_mut(bb).terminator);
+
+            for target in terminator.successors_mut() {
+                *target = final_target(mir, *target).unwrap_or(bb);
+            }
+
+            mir.basic_block_data_mut(bb).terminator = terminator;
+        }
+    }
+
+    fn simplify_branches(&self, mir: &mut Mir) -> bool {
+        let mut changed = false;
+
+        for bb in mir.all_basic_blocks() {
+            // Temporarily swap out the terminator we're modifying to keep borrowck happy
+            let mut terminator = Terminator::Diverge;
+            mem::swap(&mut terminator, &mut mir.basic_block_data_mut(bb).terminator);
+
+            mir.basic_block_data_mut(bb).terminator = match terminator {
+                Terminator::If { ref targets, .. } if targets[0] == targets[1] => {
+                    changed = true;
+                    Terminator::Goto { target: targets[0] }
+                }
+                Terminator::If { ref targets, cond: Operand::Constant(Constant {
+                    literal: Literal::Value {
+                        value: ConstVal::Bool(cond)
+                    }, ..
+                }) } => {
+                    changed = true;
+                    let target_idx = if cond { 0 } else { 1 };
+                    Terminator::Goto { target: targets[target_idx] }
+                }
+                _ => terminator
```

Comment:
@Aatch which cases do you have in mind? Handling a constant `discr` isn't easy (yet), because it's an `Lvalue` not an `Operand` (this will most likely change). I could add support for "all targets are the same" and changing a `SwitchInt` with just two targets into an `If`. Anything else you had in mind?